src/net/tcp/nbd.c

```diff
@@ -0,0 +1,1095 @@
+
+
+static inline void nbd_neg_done ( struct nbd_session *nbd ) {
+	DBGC2 ( nbd, "NBD %p negotiation done, enter transmission\n", nbd );
+	DBGC2 ( nbd, "NBD %p export size: %lld MiB, flags: 0x%04x\n", nbd,
+		nbd->export_size / 1048576, nbd->trans_flags );
+
+	nbd->rx_state = NBD_RX_TRANS_REP_CMD;
+	xfer_window_changed ( &nbd->block );
+}
+
+
+/****************************************************************************
+ *
+ * NBD to socket interface
+ *
+ */
+
+/**
+ * Pause TX engine
+ *
+ * @v nbd		NBD session
+ */
+static void nbd_tx_pause ( struct nbd_session *nbd ) {
+	process_del ( &nbd->process );
+}
+
+/**
+ * Resume TX engine
+ *
+ * @v nbd		NBD session
+ */
+static void nbd_tx_resume ( struct nbd_session *nbd ) {
+	process_add ( &nbd->process );
+}
+
+/**
+ * Start up a new transmit
+ *
+ * @v nbd		NBD session
+ * @v tx_state		New transmit state
+ *
+ * This initiates the process of sending a new data.
+ */
+static void nbd_start_tx ( struct nbd_session *nbd,
+			   enum nbd_tx_state tx_state ) {
+	assert ( nbd->tx_state == NBD_TX_IDLE );
+
+	/* Set new TX state */
+	nbd->tx_state = tx_state;
+
+	/* Start transmission process */
+	nbd_tx_resume ( nbd );
+}
+
+/**
+ * Transmit an option (NBD_OPT_EXPORT_NAME or NBD_OPT_GO)
+ *
+ * @v nbd		NBD session
+ * @ret rc		Return status code
+ */
+int nbd_tx_neg_opt ( struct nbd_session *nbd ) {
+	struct io_buffer *iobuf;
+	struct {
+		uint32_t client_flags;
+		union {
+			struct nbd_proto_opt_export_name opt_export_name;
+			struct nbd_proto_opt_go opt_go;
+		};
+	} __attribute__ (( packed )) *opt;
+	size_t name_len, opt_len;
+	int rc;
+
+	name_len = strlen ( nbd->export_name );
+	if ( nbd->use_opt_go )
+		opt_len = offsetof ( typeof ( *opt ),
+				     opt_go.export_name ) +
+				     name_len +
+				     sizeof ( uint16_t );
+	else
+		opt_len = offsetof ( typeof ( *opt ),
+				     opt_export_name.export_name ) +
+				     name_len;
+
+	iobuf = xfer_alloc_iob ( &nbd->socket, opt_len );
+	if ( ! iobuf )
+		return -ENOMEM;
+
+	opt = iob_put ( iobuf, opt_len );
+	memset ( opt, 0, opt_len );
+
+	opt->client_flags =
+		htonl ( NBD_FLAG_C_FIXED_NEWSTYLE |
+			( nbd->handshake_flags & NBD_FLAG_NO_ZEROES ?
+			NBD_FLAG_C_NO_ZEROES : 0 ) );
+
+	if ( nbd->use_opt_go ) {
+		/* Construct NBD_OPT_GO option with empty NBD_INFO list */
+		opt->opt_go.request.request_magic =
+			htonll ( NBD_OPT_REQ_MAGIC );
+		opt->opt_go.request.option = htonl ( NBD_OPT_GO );
+		opt->opt_go.request.length = htonl ( opt_len -
+			offsetof ( typeof ( *opt ), opt_go.name_length ) );
+		opt->opt_go.name_length = htonl ( name_len );
+		strcpy ( opt->opt_go.export_name, nbd->export_name );
+	} else {
+		/* Construct NBD_OPT_EXPORT_NAME option */
+		opt->opt_export_name.request.request_magic =
+			htonll ( NBD_OPT_REQ_MAGIC );
+		opt->opt_export_name.request.option =
+			htonl ( NBD_OPT_EXPORT_NAME );
+		opt->opt_export_name.request.length = htonl ( name_len );
+		strcpy ( opt->opt_export_name.export_name, nbd->export_name );
```

Comment:
Is there a guarantee that this copy cannot overrun?

Reply:
Yes, length of nbd->export_name is calculated before allocation of IOB

Reply:
OK, memcpy used